File: ai/spring/oauth/src/main/java/kr/yeotaeho/api/service/NaverOAuthService.java
package kr.yeotaeho.api.service;

import kr.yeotaeho.api.config.NaverOAuthConfig;
import kr.yeotaeho.api.dto.NaverTokenResponse;
import kr.yeotaeho.api.dto.NaverUserInfo;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.HashMap;
import java.util.Map;

/**
 * 네이버 OAuth 서비스 (RestTemplate 사용)
 * State 파라미터 검증 강화
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class NaverOAuthService {

    private final NaverOAuthConfig naverConfig;
    private final RestTemplate restTemplate;
    private final OAuthStateService stateService;

    /**
     * 네이버 로그인 URL 생성 (State 검증 지원)
     * URL 파라미터를 안전하게 인코딩하여 생성
     *
     * @return Map<String, String> {authUrl, state}
     */
    public Map<String, String> getAuthorizationUrl() {
        // State 생성 및 Redis에 저장
        String state = stateService.generateAndStoreState();

        log.info("네이버 로그인 URL 생성: state={}", state);

        try {
            // UriComponentsBuilder를 사용하여 안전하게 URL 생성
            String authUrl = UriComponentsBuilder
                    .fromUriString(NaverOAuthConfig.NAVER_AUTH_URL)
                    .queryParam("client_id", naverConfig.getClientId())
                    .queryParam("redirect_uri", naverConfig.getRedirectUri())
                    .queryParam("response_type", "code")
                    .queryParam("state", state)
                    .build()
                    .toUriString();

            log.info("생성된 네이버 인증 URL: {}", authUrl);

            Map<String, String> result = new HashMap<>();
            result.put("authUrl", authUrl);
            result.put("state", state);
            return result;

        } catch (Exception e) {
            log.error("네이버 인증 URL 생성 실패", e);
            throw new RuntimeException("네이버 인증 URL 생성 실패: " + e.getMessage());
        }
    }

    /**
     * 인가 코드로 액세스 토큰 발급
     *
     * @param code  인가 코드
     * @param state 상태 토큰
     * @return 네이버 토큰 응답
     */
    public NaverTokenResponse getAccessToken(String code, String state) {
        log.info("네이버 액세스 토큰 요청: code={}, state={}", code, state);

        log.debug("사용중인 Client ID: {}", naverConfig.getClientId());
        log.debug("사용중인 Redirect URI: {}", naverConfig.getRedirectUri());

        // HTTP 헤더 설정
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        // HTTP 바디 파라미터 설정
        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        params.add("grant_type", "authorization_code");
        params.add("client_id", naverConfig.getClientId());
        params.add("client_secret", naverConfig.getClientSecret());
        params.add("code", code);
        params.add("state", state);

        // HTTP 요청 엔티티 생성
        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(params, headers);

        try {
            // 네이버 토큰 API 호출
            ResponseEntity<NaverTokenResponse> response = restTemplate.postForEntity(
                    NaverOAuthConfig.NAVER_TOKEN_URL,
                    request,
                    NaverTokenResponse.class);

            log.info("네이버 액세스 토큰 발급 성공");
            return response.getBody();

        } catch (Exception e) {
            log.error("네이버 액세스 토큰 발급 실패: {}", e.getMessage());
            log.error("상세 에러: ", e);
            throw new RuntimeException("네이버 토큰 발급 실패: " + e.getMessage());
        }
    }

    /**
     * 액세스 토큰으로 사용자 정보 조회
     *
     * @param accessToken 액세스 토큰
     * @return 네이버 사용자 정보
     */
    public NaverUserInfo getUserInfo(String accessToken) {
        log.info("네이버 사용자 정보 조회");

        // HTTP 헤더 설정
        HttpHeaders headers = new HttpHeaders();
        headers.set("Authorization", "Bearer " + accessToken);

        // HTTP 요청 엔티티 생성
        HttpEntity<Void> request = new HttpEntity<>(headers);

        try {
            // 네이버 사용자 정보 API 호출
            ResponseEntity<NaverUserInfo> response = restTemplate.exchange(
                    NaverOAuthConfig.NAVER_USER_INFO_URL,
                    HttpMethod.GET,
                    request,
                    NaverUserInfo.class);

            NaverUserInfo userInfo = response.getBody();
            log.info("네이버 사용자 정보 조회 성공: id={}", userInfo.getResponse().getId());
            return userInfo;

        } catch (Exception e) {
            log.error("네이버 사용자 정보 조회 실패", e);
            throw new RuntimeException("네이버 사용자 정보 조회 실패: " + e.getMessage());
        }
    }

    /**
     * 전체 OAuth 플로우 실행 (State 검증 강화)
     *
     * @param code  인가 코드
     * @param state 상태 토큰
     * @return 네이버 사용자 정보
     */
    public NaverUserInfo processOAuth(String code, String state) {
        // 1. State 검증
        if (!stateService.validateAndRemoveState(state)) {
            throw new RuntimeException("State 검증 실패: CSRF 공격 가능성");
        }

        // 2. 토큰 발급
        NaverTokenResponse tokenResponse = getAccessToken(code, state);

        // 3. 사용자 정보 조회
        return getUserInfo(tokenResponse.getAccessToken());
    }
}

File: backend/domain/oauth/service/naver_oauth_service.py
import httpx
from typing import Dict, Any, Optional
import logging
from urllib.parse import urlencode
from ..config.settings import settings
from ..util.state import OAuthStateService

logger = logging.getLogger(__name__)


class NaverOAuthService:
    """네이버 OAuth 서비스"""
    
    NAVER_AUTH_URL = "https://nid.naver.com/oauth2.0/authorize"
    NAVER_TOKEN_URL = "https://nid.naver.com/oauth2.0/token"
    NAVER_USER_INFO_URL = "https://openapi.naver.com/v1/nid/me"
    
    def __init__(
        self,
        state_service: OAuthStateService,
        http_client: httpx.AsyncClient
    ):
        self.state_service = state_service
        self.http_client = http_client
        self.client_id = settings.naver_client_id
        self.client_secret = settings.naver_client_secret
        self.redirect_uri = settings.naver_redirect_uri
    
    async def get_authorization_url(self, mode: Optional[str] = None) -> Dict[str, str]:
        """네이버 로그인 URL 생성 (State 검증 지원)"""
        # State 생성 및 Redis에 저장 (mode 정보 포함)
        state = await self.state_service.generate_and_store_state(mode=mode)
        
        logger.info(f"네이버 로그인 URL 생성: state={state}")
        
        # Authorization URL 생성
        params = {
            "client_id": self.client_id,
            "redirect_uri": self.redirect_uri,
            "response_type": "code",
            "state": state
        }
        
        auth_url = f"{self.NAVER_AUTH_URL}?{urlencode(params)}"
        
        logger.info(f"생성된 네이버 인증 URL: {auth_url}")
        
        return {
            "authUrl": auth_url,
            "state": state
        }
    
    async def get_access_token(self, code: str, state: str) -> Dict[str, Any]:
        """인가 코드로 액세스 토큰 발급"""
        logger.info(f"네이버 액세스 토큰 요청: code={code}, state={state}")
        
        # 토큰 요청 파라미터
        data = {
            "grant_type": "authorization_code",
            "client_id": self.client_id,
            "client_secret": self.client_secret,
            "code": code,
            "state": state
        }
        
        try:
            response = await self.http_client.post(
                self.NAVER_TOKEN_URL,
                data=data,
                headers={"Content-Type": "application/x-www-form-urlencoded"}
            )
            response.raise_for_status()
            
            token_data = response.json()
            logger.info("네이버 액세스 토큰 발급 성공")
            return token_data
            
        except httpx.HTTPError as e:
            logger.error(f"네이버 액세스 토큰 발급 실패: {e}")
            raise RuntimeError(f"네이버 토큰 발급 실패: {e}")
    
    async def get_user_info(self, access_token: str) -> Dict[str, Any]:
        """액세스 토큰으로 사용자 정보 조회"""
        logger.info("네이버 사용자 정보 조회")
        
        try:
            response = await self.http_client.get(
                self.NAVER_USER_INFO_URL,
                headers={"Authorization": f"Bearer {access_token}"}
            )
            response.raise_for_status()
            
            user_info = response.json()
            logger.info(f"네이버 사용자 정보 조회 성공: id={user_info.get('response', {}).get('id')}")
            return user_info
            
        except httpx.HTTPError as e:
            logger.error(f"네이버 사용자 정보 조회 실패: {e}")
            raise RuntimeError(f"네이버 사용자 정보 조회 실패: {e}")
    
    async def process_oauth(self, code: str, state: str) -> Dict[str, Any]:
        """전체 OAuth 플로우 실행 (State 검증 강화)"""
        # 1. State 검증 및 mode 정보 추출
        state_data = await self.state_service.validate_and_remove_state(state)
        if not state_data:
            raise RuntimeError("State 검증 실패: CSRF 공격 가능성")
        
        mode = state_data.get("mode")
        
        # 2. 토큰 발급
        token_response = await self.get_access_token(code, state)
        
        # 3. 사용자 정보 조회
        user_info = await self.get_user_info(token_response["access_token"])
        
        # mode 정보를 user_info에 추가
        if mode:
            user_info["_mode"] = mode
        
        return user_info

